### PR TITLE
Enhanced rich text and encoding/decoding fixes

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

This PR aims to fixes multiple issues:
 - encoding/decoding issues that may alter html special chars (#140, #2247),
 - newlines alteration that prevent correct usage of preformatted text (#5607, #8757),
 - bad rendering of content that has been stored as plain text (#5379),
 - rich text rendering in tooltips (#7981)
 - ...

### CODE REVIEW

**Comment by orthagh** [comment]
Adding code sample plugin is a good addition, a question by the way, is copy pasting directly a xml into textarea results in a non altered text with this pr, now ?

**Comment by cedric-anne** [comment]
> Adding code sample plugin is a good addition, a question by the way, is copy pasting directly a xml into textarea results in a non altered text with this pr, now ?

Yes, pasting XML will the textarea will works as expected :
image

File: install/migrations/update_0.84.0_to_0.84.1.php [hunk 1]
```diff
@@ -72,7 +72,7 @@ function update0840to0841() {
                             'glpi_notificationtemplatetranslations' => 'content_text'];
    foreach ($fields_to_clean as $table => $field) {
       foreach ($DB->request($table) as $data) {
-         $text  = Toolbox::unclean_html_cross_side_scripting_deep($data[$field]);
+         $text  = @Toolbox::unclean_html_cross_side_scripting_deep($data[$field]);
```
Comment (trasher) [inline install/migrations/update_0.84.0_to_0.84.1.php:75]: I do not think this is a good idea to mute errorrs; this one will have to be rewritten aswell, doesn't it?

**Review by trasher — COMMENTED** [review]
Seems ok from a technical point of view; even if it's very difficult to follow all changes and "predict" the results (but I guess you're already aware of that ;p).

Many parts of GLPI may be affected, I'll do some functional tests on my side today.

File: install/migrations/update_0.84.0_to_0.84.1.php (hunk 1 above)
Comment (cedric-anne) [reply install/migrations/update_0.84.0_to_0.84.1.php:75]: Maybe this whole loop can be dropped, but I have to make manual tests on this.

This has been done to fix https://forum.glpi-project.org/viewtopic.php?id=32196 in commit f19be70f24d0686a4477923fed0b6fc60782f901 .

If I manually put a value corresponding to thoose provided in forum thread, it is correctly displayed without having to run this migration.

Comment (cedric-anne) [reply install/migrations/update_0.84.0_to_0.84.1.php]: I made some tests and I think we could drop this part (done in my last commit).

Indeed, the only purpose was to re-encode HTML entities from a numeric encoding to the default encoding (e.g. from `&#233;` to `&eacute;`). As both encoding are valid HTML encodings, I do not get why it has been done like this.

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**